pkg/gardenlet/bootstrap/certificate/certificate_rotation.go

```diff
@@ -0,0 +1,233 @@
+// limitations under the License.
+
+package certificate
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/gardener/gardener/pkg/gardenlet/bootstrap/util"
+	"github.com/gardener/gardener/pkg/logger"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/retry"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// certificateWaitTimeout controls the amount of time we wait for the certificate
+// approval in one iteration.
+var certificateWaitTimeout = 15 * time.Minute
+
+// Manager can be used to schedule the certificate rotation for the Gardenlet's Garden cluster client certificate
+type Manager struct {
+	stopCh chan struct{}
+
+	seedName               string
+	clientMap              clientmap.ClientMap
+	gardenClientConnection config.GardenClientConnection
+
+	logger logrus.FieldLogger
+	// Set to time.Now but can be stubbed out for testing
+	now func() time.Time
+}
+
+// NewCertificateRotation creates a certificate manager that can be used to rotate gardenlet's client certificate for the Garden cluster
+func NewCertificateManager(gardenClientConnection config.GardenClientConnection, seedConfig *config.SeedConfig, clientMap clientmap.ClientMap) *Manager {
+	seedName := util.GetSeedName(seedConfig)
+	return &Manager{
+		stopCh:                 make(chan struct{}),
+		seedName:               seedName,
+		clientMap:              clientMap,
+		gardenClientConnection: gardenClientConnection,
+		logger:                 logger.NewFieldLogger(logger.Logger, "certificate-rotation", seedName).Logger,
+		now:                    time.Now,
+	}
+}
+
+// Stop stops the certificate rotation
+func (cr *Manager) Stop() {
+	close(cr.stopCh)
+}
+
+// ScheduleCertificateRotation waits until the currently used Garden cluster client certificate approaches expiration.
+// Then requests a new certificate and stores the kubeconfig in a secret (`gardenClientConnection.kubeconfigSecret`) on the Seed.
+// the argument is context.Cancle function to cancle the context of the Gardenlet used for graceful termination after a successful certificate rotation.
+// when the new gardenlet pod is started, it uses the rotated certificate stored in the secret in the Seed cluster
+func (cr *Manager) ScheduleCertificateRotation(gardenletCancle context.CancelFunc) {
+	go wait.Until(func() {
+		if err := cr.waitForCertificateRotation(); err != nil {
+			cr.logger.Errorf("waiting for the certificate rotation failed: %v", err)
+			return
+		}
+
+		err := retry.Until(context.TODO(), certificateWaitTimeout, func(ctx context.Context) (bool, error) {
```

Comment:
```suggestion
		err := retry.Until(ctx, certificateWaitTimeout, func(ctx context.Context) (bool, error) {
```

Out of curiosity: Why is it retried, i.e., what would happen if we don't retry and it fails? What happens now if it fails for the configured `15m`?

Reply:
It is to make sure that sending the CSR is retried.
If we do not retry, the certificate might not get rotated.
If it fails within 15 minutes, it is retried as of now.

This is also how the kubelet does it.